concurrency/ParallelTransformingIterable.java

```diff
@@ -0,0 +1,141 @@
+    @NonNull
+    private final Iterable<S> source;
+    @NonNull
+    private final Function<S, T> transformer;
+    @NonNull
+    private final ExecutorService executor;
+    @NonNull
+    private final Integer parallelism;
+
+    @Override
+    public Iterator<T> iterator() {
+        return new ParallelTransformingIterator(source, transformer, executor, parallelism);
+    }
+
+    /*
+     * Main iterator that initializes the multi-segment iterables and maintains the blocking queue.
+     */
+    private final class ParallelTransformingIterator extends AbstractIterator<T> {
+        private static final int POLL_TIMEOUT_MILLIS = 10;
+        private static final int PROGRESS_MILESTONE = 1000;
+
+        private final Iterable<S> source;
+        private final Function<S, T> transformer;
+        private final Integer parallelism;
+
+        private final ConcurrentLinkedQueue<T> queue = new ConcurrentLinkedQueue<T>();
+        private final ListeningExecutorService listeningExecutor;
+        private final Iterator<List<S>> sourceIt;
+
+        private final AtomicBoolean done = new AtomicBoolean(false);
+        private final AtomicInteger transformationFailures = new AtomicInteger(0);
+        private final AtomicInteger inflightTransformations = new AtomicInteger(0);
+        private final AtomicLong iteratedFromSource = new AtomicLong(0);
+
+        /**
+         * Private to ParallelTransformingIterable which can not pass null variables.
+         * 
+         * @param source
+         * @param transformer
+         * @param executor
+         * @param parallelism
+         */
+        ParallelTransformingIterator(final Iterable<S> source, final Function<S, T> transformer, final ExecutorService executor,
+                final Integer parallelism) {
+            this.source = source;
+            this.transformer = transformer;
+            this.listeningExecutor = MoreExecutors.listeningDecorator(executor);
+            this.parallelism = parallelism;
+            this.sourceIt = Iterables.partition(source, parallelism).iterator();
+        }
+
+        @Override
+        protected T computeNext() {
+            if (transformationFailures.get() > 0) {
+                throw new RuntimeException("Parallel iteration failed. Some source elements could not be transformed");
+            }
+
+            if (inflightTransformations.get() < parallelism) {
+                submitNextBatch();
+            }
+
+            if (inflightTransformations.get() == 0 && queue.size() == 0) {
+                log.info("Iterated from source {} ", iteratedFromSource.get());
+                return endOfData();
+            }
+
+            while (queue.size() == 0) {
+                try {
+                    Thread.sleep(POLL_TIMEOUT_MILLIS);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+
+            return queue.poll();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentLinkedQueue`, but your usage of `size()` and `poll()` may not be thread-safe at lines: **98 and 106**. If the `ConcurrentLinkedQueue` contains only **one element** and two threads perform this **same check at the same time** , the second thread&#39;s call to `poll()` will get `null`. This `null` may be used at line: **106**.

**Fix**  
 Consider calling `poll()`, checking instead of your current check if the returned object is `null`, and only then using that object.